consensus/spos/bls/subroundSignature.go

```diff
@@ -378,59 +380,100 @@ func (sr *subroundSignature) remainingTime() time.Duration {
 	return remainigTime
 }
 
+// MultikeySigData and its fields are exported to be accessible in tests
+type MultikeySigData struct {
+	IsCurrentNodeMultiKeyLeader bool
+	IsFlagActive                bool
+	Mutex                       *sync.Mutex
+	NumMultiKeysSignaturesSent  *int
+}
+
 func (sr *subroundSignature) doSignatureJobForManagedKeys() bool {
 	isCurrentNodeMultiKeyLeader := sr.IsMultiKeyLeaderInCurrentRound()
 	isFlagActive := sr.EnableEpochsHandler().IsFlagEnabledInEpoch(common.EquivalentMessagesFlag, sr.Header.GetEpoch())
 
+	mutex := sync.Mutex{}
 	numMultiKeysSignaturesSent := 0
-	for idx, pk := range sr.ConsensusGroup() {
-		pkBytes := []byte(pk)
-		if sr.IsJobDone(pk, sr.Current()) {
-			continue
-		}
-		if !sr.IsKeyManagedByCurrentNode(pkBytes) {
-			continue
-		}
 
-		signatureShare, err := sr.SigningHandler().CreateSignatureShareForPublicKey(
-			sr.GetData(),
-			uint16(idx),
-			sr.Header.GetEpoch(),
-			pkBytes,
-		)
-		if err != nil {
-			log.Debug("doSignatureJobForManagedKeys.CreateSignatureShareForPublicKey", "error", err.Error())
-			return false
-		}
+	multiKeySigData := MultikeySigData{
+		IsCurrentNodeMultiKeyLeader: isCurrentNodeMultiKeyLeader,
+		IsFlagActive:                isFlagActive,
+		Mutex:                       &mutex,
+		NumMultiKeysSignaturesSent:  &numMultiKeysSignaturesSent,
+	}
+
+	sentSigForAllKeys := true
+	wg := sync.WaitGroup{}
+	wg.Add(len(sr.ConsensusGroup()))
 
-		isCurrentManagedKeyLeader := idx == spos.IndexOfLeaderInConsensusGroup
-		// TODO[cleanup cns finality]: update the check
-		// with the equivalent messages feature on, signatures from all managed keys must be broadcast, as the aggregation is done by any participant
-		shouldBroadcastSignatureShare := (!isCurrentNodeMultiKeyLeader && !isFlagActive) ||
-			(!isCurrentManagedKeyLeader && isFlagActive)
-		if shouldBroadcastSignatureShare {
-			ok := sr.createAndSendSignatureMessage(signatureShare, pkBytes)
-			if !ok {
-				return false
+	for idx, pk := range sr.ConsensusGroup() {
```

Comment:
it might be better to determine first which keys are managed by current node, and try to send signatures only for those, instead of creating a separate goroutine for each key in consensus group

Reply:
maybe add a TODO comment for this

Reply:
updated with a TODO